rfcs/45-stageless.md

```diff
@@ -0,0 +1,537 @@
+
+The complete inability to configure plugins is the source of [significant user pain](https://github.com/bevyengine/bevy/issues/2160).
+Plugin authors cannot know all of the needs of the end user's app, and so some degree of control must be handed back to the end users.
+
+This control is carefully limited though: systems can only be configured if they are given a public label, and all of the systems under a common label are configured as a group.
+In addition, configuration defined by a private label (or directly on systems) cannot be altered, extended or removed by the app: it cannot be named, and so cannot be altered.
+
+These limitation allows plugin authors to carefully design a public API, ensure critical invariants hold, and make serious internal changes without breaking their dependencies (as long as the public labels and their meanings are relatively stable).
+
+### Why can't you label labels?
+
+If each label is associated with a `SystemConfig`, and part of this struct is which labels are applied, why can't we label labels themselves?
+
+In short: this way lies madness.
+While at first it may seem appealing to be able create nested hierarchies of labels, this ends up recreating many of the problems of a stage-based design's global perspective.
+If labels can themselves cause other labels to be applied, it becomes dramatically harder to track how systems are configured, and why they are configured in that way.
+
+This is a problem for users, as it makes code much harder to read, reason about and maintain.
+It is also a problem for our tools, forcing a recursive architecture when generating schedules and limiting the ability for developer tools to communicate both the how and why of a system's configuration.
+
+### Why is if-needed the correct default ordering strategy?
+
+Strict ordering is simple and explicit, and will never result in strange logic errors.
+On the other hand, it *will* result in pointless and surprising blocking behavior, possibly leading to unsatisfiable schedules.
+
+If-needed ordering is the corret stategy in virtually cases: in Bevy, interior mutability at the component or resource level is rare, almost never needed and results in other serious and subtle bugs.
+As we move towards specifying system ordering dependencies at scale, it is critical to avoid spuriously breaking users schedules, and silent, pointless performance hits are never good.
+
+### Why can't we just use system chaining for run criteria?
+
+There are two reasons why this doesn't work:
+
+1. The system we're applying a run criteria does not have an input type.
+2. System chaining does not work if the chained systems are incompatible. This is far too limiting.
+
+### Why aren't run criteria cached?
+
+In practice, almost all run criteria are extremely simple, and fast to check.
+Verifying that the cache is still valid will require access to the data anyways, and involve more overhead than simple computations on one or two resources.

```

Comment:
Just to be clear: it is still possible to have a run criteria only run once, but control the execution of many systems, right?

(i assume this would be done by putting a run criteria on a label, affecting all systems with that label)

Reply:
>Just to be clear: it is still possible to have a run criteria only run once, but control the execution of many systems, right?

No. As of the first draft, I explicitly decided against that behavior, because it induces a lot of atomic order constraints. Instead, adding run criteria via a label inserts multiple copies of the run criteria system, each evaluated separately.

Critically, those ordering constraints would cause schedules which have the same run criteria (added by a label) to result in an unsatisfiable schedule if there are exclusive systems in between. This effectively gets us back to the frustrations of "why can't I use this across stages".

The foot-gun is not worth the questionable performance benefits here. If you really need to cache expensive results, do it manually.

Reply:
I don't think the performance benefits are "questionable".

It's not about whether the results/computations are expensive. Precisely the opposite. They are often so cheap, that "executing a system" is large non-negligible overhead in comparison.

I know we often like to consider "executing a system" to be a "fast" operation, and that users "shouldn't worry about the cost of running many small systems". And it might be, if you are comparing it to the cost of iterating entities and performing algorithmic work, which is what most user systems do. But not run criteria.

Executing a system requires at least an indirect function call, often with the overhead of multithreaded scheduling (dispatch onto the thread pool), and memory allocation. All of these are *very* expensive operations, compared to just checking a value (the body of the typical run criteria).

That's *a lot* of extra overhead, to simply check a resource value for equality (the most common run criteria, for example, states).

Also, with your proposed design, this is overhead that *there is nothing the user can do to mitigate!* Even if they make the body of the function as trivial as possible, even if it is just a `return true;`, they must still pay all of this cost, for every system they run.

It effectively turns every system execution into multiple system executions. Every system gated behind a state (or any run criteria), which would be almost all user systems in a typical large project, would be paying this cost. Adding more systems to the schedule effectively becomes at least twice as expensive in terms of overhead, often more.

I really worry that your proposed design would introduce a lot of hidden, mandatory, performance cost.

Reply:
Honestly, I'd be happy, if we special-case / hardcode the proposed trivial `if_resource_equals` run criteria (which is what states and stuff would be based on), into the executor/scheduler implementation, so that it is not really a system/function like other run criteria, but the check is just done by the scheduler directly inline, saving all the runtime overhead in the most common and trivial use case.

That would at least mitigate the cost where it matters most.